nearswap/src/twap.rs

```diff
@@ -0,0 +1,553 @@
+    pub fn write(
+        &mut self,
+        block_timestamp: u64,
+        price1: u128,
+        price2: u128,
+        max_length: usize
+    ) -> usize {
+        let last: &Observation = &self.observations[self.last_updated_index].clone();
+        if(block_timestamp == u64::try_from(last.block_timestamp).unwrap()) {
+            self.observations[self.last_updated_index] = Observation::transform(last, block_timestamp, price1, price2);
+            return self.last_updated_index;
+        }
+
+        if self.last_updated_index + 1 >= max_length {
+            self.pivoted = true;
+        }
+
+        let updated_index: usize = (self.last_updated_index + 1) % max_length;
+        if updated_index < self.observations.len() {
+            self.observations[updated_index] = Observation::transform(last, block_timestamp, price1, price2);
+        } else {
+            self.observations.push(Observation::transform(last, block_timestamp, price1, price2));
+        }
+
+        self.last_updated_index = updated_index;
+        return updated_index;
+    }
+
+    /**
+    Pivoted point binary search: searches array which is
+    sorted and rotated from a particular point.
+    Similar to rotated array from a certain pivot point.
+    Parameters:
+    + `max_length`: The maximum length of TWAP array.
+    + `timestamp`: given timestamp.
+    */
+    pub fn binary_search(
+        &self,
+        max_length: usize,
+        block_timestamp: u64,
+    ) -> usize {
+
+        // edge case when all values are less than required
+        if u64::try_from(self.observations[self.last_updated_index].block_timestamp).unwrap() 
+            < u64::try_from(block_timestamp).unwrap() {
+            panic!("Observation after this timestamp doesn't exist");
+        }
+
+        let mut start: usize = 0;
+        let mut end: usize = self.last_updated_index + 1;
+
+        let mut mid: usize;
+
+        while start < end {
+            mid = (start + end) / 2;
+            if u64::try_from(block_timestamp).unwrap() <= u64::try_from(self.observations[mid].block_timestamp).unwrap() {
+                end = mid;
+            } else {
+                start = mid + 1;
+            }
+        }
+
+        if self.pivoted && start == 0 {
+            let res = start;
+            start = self.last_updated_index + 1;
+            end = max_length;
+
+            while start < end {
+                mid = (start + end) / 2;
+                if u64::try_from(block_timestamp).unwrap() <= u64::try_from(self.observations[mid].block_timestamp).unwrap() {
+                    end = mid;
+                } else {
+                    start = mid + 1;
+                }
+            }
+            if start == max_length {
+                start = res;
+            }
+
+            return start;
+        }
+
+        return start;
+    }
+
+    // convert seconds into nanoseconds
+    pub fn to_nanoseconds(&self, time: u64) -> u64 {
+        return time * 1000_000_000;
+    } 
```

Comment:
this should be module function, not a class function.

Reply:
ok